src/validation.ts

```diff
@@ -0,0 +1,137 @@
+import { Client, cacheExchange, fetchExchange } from '@urql/core';
+import {
+  type DscvrFramesRequest,
+  type DscvrValidationResponse,
+  dscvrClientProtocolPrefix,
+  DscvrUntrustedData,
+} from './types';
+import { DEFAULT_DSCVR_API_URL } from './default';
+
+type ValidationType = string | number | bigint | null | undefined;
+interface ValidatedQueryResult {
+  user: {
+    id: string;
+  };
+  content?: {
+    id: string;
+  };
+  state: string | null;
+  inputText: string | null;
+  timestamp: string;
+  url: string;
+  buttonIndex: number;
+}
+
+const validationQuery = `#graphql
+    query validate($message: String!) {
+      unpackFrameMessage(message: $message) {
+        user {
+            id
+        },
+        content {
+            id
+        }
+        buttonIndex,
+        state,
+        url,
+        timestamp,
+        inputText,
+      }
+    }
+`;
+
+export const validateClientProtocol = (clientProtocol: string) => {
+  return clientProtocol.startsWith(dscvrClientProtocolPrefix);
+};
+
+const validateUntrustedDataValue = (
+  validated: ValidationType,
+  untrusted: ValidationType
+): boolean => {
+  if (!validated && !untrusted) {
+    return true;
+  }
+  if (validated && untrusted && validated === untrusted) {
+    return true;
+  }
+  return false;
+};
+
+const validateUntrustedData = (
+  validatedResult: DscvrValidationResponse,
+  untrustedData: DscvrUntrustedData
+): boolean => {
+  if (
+    validateUntrustedDataValue(
+      validatedResult.validatedDscvrId,
+      untrustedData.dscvrId
+    ) &&
+    validateUntrustedDataValue(
```

Comment:
If you follow my suggestion above, then you can just loop through the superset of the keys in the trusted and untrusted data

Reply:
Hmm, thats tricky, TS does not have a way to loop through the superset of a type, only for an object, and optional props wont appear in that loop if undefined.

The way to do this with a loop is this, but I honestly prefer the explicit if, WDYT?

```
  const keys: (keyof DscvrValidationResponse)[] = 
  [
    'dscvrId',
    'contentId',
    'buttonIndex',
    'url',
    'inputText',
    'state',
    'timestamp',
  ];
  keys.reduce((valid, key) => {
    return valid && validateField(validatedResult[key], untrustedDataChanged[key]);
  }, true);
```

Reply:
The only downside to this is the untrusted data can have fields that are not in the validated response and we wouldn't catch them.

Not to create more work, but how about setting any fields that are null in `validateField` to undefined, and then you can make sure the number of keys in untrusted is the same as the valid ones and also loop through all the keys and make sure they match up. Wdyt?

Reply:
Omg, you are absolutely right, I didn't consider the fact that untrusted data could have extra attributes
thanks I changed the code